File: java-index-checker/src/main/java/org/example/ic9/ArrayWrapper.java
package org.example.ic9;

import org.checkerframework.checker.index.qual.SameLen;
import org.checkerframework.checker.index.qual.IndexFor;
import org.checkerframework.checker.index.qual.NonNegative;
import org.checkerframework.checker.index.qual.LengthOf;


/** ArrayWrapper is a fixed-size generic collection. */
public class ArrayWrapper<T> {
    private final Object @SameLen("this") [] delegate;

    @SuppressWarnings("index") // constructor creates object of size @SameLen(this) by definition
    ArrayWrapper(@NonNegative int size) {
        delegate = new Object[size];
    }

    public @LengthOf("this") int size() {
        return delegate.length;
    }

    public void set(@IndexFor("this") int index, T obj) {
        delegate[index] = obj;
    }

    @SuppressWarnings("unchecked") // required for normal Java compilation due to unchecked cast
    public T get(@IndexFor("this") int index) {
        return (T) delegate[index];
    }
}

File: scala/src/main/scala/ic9/ArrayWrapper.scala
package ic9


class ArrayWrapper[T] private(private val delegate: Array[Any]) {

  def this(size: Int) =
    this(new Array[Any](size))    //> ArrayWrapper::constructor aux-annot=1

  export delegate.length as size  //> ArrayWrapper::size

  export delegate.update as set   //> ArrayWrapper::set
  
  export delegate.apply as get    //> ArrayWrapper::get

}
